src/pass/onnx_rewrite.td

```diff
@@ -24,4 +24,54 @@ include "dialect/onnx/onnx.td"
 ///    dag benefitsAdded = (addBenefit 0)
 /// >;
 
+// Create a StringAttr from a string.
+class StringAttrOfValue<string val>: NativeCodeCall<"$_builder.getStringAttr(\"" # val # "\")">;
+
+// Create a FloatAttr from an interger value.
+// It seems Table-gen does not support `float` type, so we can not pass a float value.
+class FloatAttrOfValue<int val> : NativeCodeCall<"FloatAttr::get($0.getType().cast<TensorType>().getElementType(), " # val # ")">;
+
+// Create an ArrayAttr of IntergerAttr(s) of zero values.
+// This function is used for padding attribute in MaxPoolSingleOut.
+class ArrayAttrOfZeroWithTrail<int trailCount> : NativeCodeCall<"createArrayAttrOfZeroWithTrail($_builder, $0, " # trailCount # ")">;
+
+// Pad a ArrayAttr with trails of zeros.
+// This function is used for padding attribute in MaxPoolSingleOut.
+class PadArrayAttrWithZeroTrail<int trailCount> : NativeCodeCall<"padArrayAttrWithZeroTrail($_builder, $0," # trailCount # ")">;
+
+// Check whether an ArrayAttr contains non-zero values or not.
+def HasNonZeroInArrayAttr : Constraint<CPred<"hasNonZeroInArrayAttr($_self)">, "has non-zero elements">;
+def HasAllZeroInArrayAttr : Constraint<CPred<"!hasNonZeroInArrayAttr($_self)">, "has non-zero elements">;
+
+//===----------------------------------------------------------------------===//
+// Rewrite:
+// %0 = onnx.MaxPoolSingleOutOp(%D : tensor<DShape>)
+//     {pads = [b0, b1, ... bK, e0, e1, ..., eK]} ->
+//         tensor<OutShape>
+//
+// as:
+// %0 = onnx.PadConstantValuePadOp(%D)
+//     {pads = [0, 0, b0, b1, ... bK, 0, 0, e0, e1, ..., eK]} ->
+//     tensor<DPaddedShape>
+// %1 = onnx.MaxPoolSingleOut(%0 : tensor<DPaddedShape>) {pads = [0, ..., 0]} ->
+//     tensor<OutShape>
+//===----------------------------------------------------------------------===//
+
+def MaxPoolSingleOutOpPaddingPattern: Pat<
+  (ONNXMaxPoolSingleOutOp:$res
+     $x,
+     $auto_pad, $ceil_mode, $dilation, $kernel_shape,
+     $pads,
+     $storage_order, $strides),
+  (ONNXMaxPoolSingleOutOp
+     (ONNXPadConstantValuePadOp $x,
+        (PadArrayAttrWithZeroTrail<2> $pads),
+        (FloatAttrOfValue<0> $res),
```

Comment:
Does this $res have anything to do with the $res at line 61? If not can we name it differently? The name of this attribute in the definition is constant_value so maybe we can give it the same name.